lib/ts/test_MT_hashtable.cc

```diff
@@ -0,0 +1,175 @@
+      fprintf(stderr, ".");
+    }
+    removed_count++;
+  }
+  fprintf(stderr, "done\n");
+
+  fprintf(stdout, "%ld data entries are removed\n", removed_count);
+  fprintf(stdout, "verify the content again\n");
+  for (i = 1; i <= count; i++) {
+    long data = htable->lookup_entry(i);
+    if (i % 2 == 1 && data == 0) {
+      fprintf(stdout, "verify content failed: key(%ld) deleted\n", i);
+      *pstatus = TEST_FAILED;
+      delete htable;
+      return;
+    }
+    if (data != 0 && data != i) {
+      fprintf(stdout, "verify content failed: key(%ld) data(%ld)\n", i, data);
+      *pstatus = TEST_FAILED;
+      delete htable;
+      return;
+    }
+    if (i % (count / 50) == 0) {
+      fprintf(stderr, ".");
+    }
+  }
+  fprintf(stderr, "done\n");
+
+  fprintf(stdout, "use iterator to list all the elements and delete half of them\n");
+  HashTableIteratorState<long, long> it;
+  int j, new_count = 0;
+  for (j = 0; j < MT_HASHTABLE_PARTITIONS; j++) {
+    int data = htable->first_entry(j, &it);
+    while (data > 0) {
+      new_count++;
+      if (new_count % (count / 25) == 0) {
+        fprintf(stderr, ".");
+      }
+
+      if (new_count % 2 == 0) {
+        htable->remove_entry(j, &it);
+        data = htable->cur_entry(j, &it);
+        removed_count++;
+      } else {
+        data = htable->next_entry(j, &it);
+      }
+    }
+  }
+  fprintf(stderr, "done\n");
+
+  fprintf(stdout, "verify the content once again\n");
+  new_count = count - removed_count;
+  for (j = 0; j < MT_HASHTABLE_PARTITIONS; j++) {
+    int data = htable->first_entry(j, &it);
+    while (data > 0) {
+      new_count--;
+      if (new_count % (count / 25) == 0) {
+        fprintf(stderr, ".");
+      }
+      data = htable->next_entry(j, &it);
+      if (data != htable->lookup_entry(data)) {
+        fprintf(stdout, "verify content failed: key(%d) data(%ld)\n", data, htable->lookup_entry(data));
+        *pstatus = TEST_FAILED;
+        delete htable;
+        return;
+      }
+    }
+  }
+
+  fprintf(stderr, "done\n");
+  if (new_count != 0) {
+    fprintf(stdout, "there are %d extra entries in the table\n", new_count);
+    *pstatus = TEST_FAILED;
+    delete htable;
+    return;
+  }
+
+  fprintf(stdout, "remove everything using iterator\n");
+  new_count = count - removed_count;
+  for (j = 0; j < MT_HASHTABLE_PARTITIONS; j++) {
+    int data = htable->first_entry(j, &it);
+    while (data > 0) {
+      new_count--;
+      if (new_count % (count / 25) == 0) {
+        fprintf(stderr, ".");
+      }
+      htable->remove_entry(j, &it);
+      data = htable->cur_entry(j, &it);
+    }
+  }
+
+  fprintf(stderr, "done\n");
+  if (new_count != 0) {
+    fprintf(stdout, "there are %d extra entries in the table\n", new_count);
+    *pstatus = TEST_FAILED;
+    delete htable;
+    return;
+  }
+
+  delete htable;
+  *pstatus = TEST_PASSED;
+}
+
+int
+main(int /* argc ATS_UNUSED */, char ** /*argv ATS_UNUSED */)
+{
+  int ret;
+  test_MT_HashTable(&ret);
```

Comment:
Please use "Catch" for unit test framework. You can find some tests using Catch in `lib/ts/unit-tests/`
https://github.com/apache/trafficserver/tree/master/lib/ts/unit-tests

The header file is https://github.com/apache/trafficserver/blob/master/tests/include/catch.hpp
Docs are https://github.com/catchorg/Catch2/blob/master/docs/Readme.md#top